aptos-move/move-examples/nfts_as_accounts/sources/token.move

```diff
@@ -0,0 +1,232 @@
+///    * Additional token metadata can also be stored as other resources in this account.
+///    * When creating a Token, a TokenRef is returned for managing that Token.
+///    * The TokenRef is what offers composability of Tokens, since a Token can contain other
+///       TokenRefs.
+///    * The creator continues to be the "manager" of the token even after creation.
+///    * A Token is globally accessible and cannot be hidden, thus a creator always has access to
+///       the token to mutate it as appropriate.
+///
+/// Some notes:
+/// * This module doesn't have clean error handling because some operations are not cheap. It is
+///   more effective to just let the upstream abort happen, rather than try to handle it locally,
+///   unless there's a means to recover.
+///
+/// TODO(@davidiw):
+/// * add appropriate set of accessors for fields
+/// * consider tracking ownership if the token is inserted into a TokenStore
+/// * create_token should probably have another function that returns the signer for descendants
+module nfts_as_accounts::token {
+    use std::option::{Option, Self};
+    use std::signer;
+    use std::string::{Self, String};
+    use std::vector;
+
+    use aptos_framework::account;
+
+    const ENOT_THE_CREATOR: u64 = 1;
+    const EINVALID_REF_CONVERSION: u64 = 2;
+
+    /// Represents the common fields to all tokens.
+    struct Token<Data> has key {
+        /// An optional categorization of similar token, there are no constraints on collections.
+        collection: String,
+        /// The original creator of this token.
+        creator: address,
+        /// A brief description of the token.
+        description: String,
+        /// Determines which fields are mutable.
+        mutability_config: MutabilityConfig,
+        /// The name of the token, which should be unique within the collection; the length of name
+        ///should be smaller than 128, characters, eg: "Aptos Animal #1234"
+        name: String,
+        /// The denominator and numerator for calculating the royalty fee; it also contains payee
+        /// account address for depositing the Royalty
+        royalty: Royalty,
+        /// The Uniform Resource Identifier (uri) pointing to the JSON file stored in off-chain
+        /// storage; the URL length will likely need a maximum any suggestions?
+        uri: String,
+        /// Inner data
+        data: Option<Data>,
+    }
+
+    /// This config specifies which fields in the TokenData are mutable
+    struct MutabilityConfig has copy, drop, store {
+        description: bool,
+        name: bool,
+        uri: bool,
+    }
+
+    /// The royalty of a token
+    struct Royalty has copy, drop, store {
+        numerator: u64,
+        denominator: u64,
+        /// The recipient of royalty payments. See the `shared_account` for how to handle multiple
+        /// creators.
+        payee_address: address,
+    }
+
+    struct TokenRef<phantom Data> has store {
```

Comment:
Lets say you have a `TokenRef<BaseToken>`. You do not know the actual type of the data, but you like to access common info like `creator`. How? It seems not possible because you can only access storage if you know the concrete `Data` (?)

Reply:
With other words, here the `Any` based solution is strictly superior. You can access generic data without knowning concrete one. @davidiw

Reply:
I can see a lot of value in that direction. I think there's a second level of work that needs to be done here that evaluates performance of the various implementations. In this case very much focused on read / writes.